### PR TITLE
.github/workflows: add integration test with a custom runner

### PR DESCRIPTION
**Pull request by Xe**
This runner is in my homelab while we muse about a better, more
permanent home for these tests.

Updates #1988

Signed-off-by: Christine Dodrill <xe@tailscale.com>

### CODE REVIEW

File: tstest/integration/vms/vms_test.go [hunk 1]
```diff
@@ -294,6 +294,8 @@ func mkVM(t *testing.T, n int, d Distro, sshKey, hostURL, tdir string) func() {
 	t.Logf("running: qemu-system-x86_64 %s", strings.Join(args, " "))
 
 	cmd := exec.Command("qemu-system-x86_64", args...)
+	cmd.Stdout = log.Writer()
```
Comment (bradfitz) [inline tstest/integration/vms/vms_test.go:297]: let's not change this file in this commit/PR.  These are two unrelated things. They should have their own commits.

File: .github/workflows/linux.yml [hunk 2]
```diff
@@ -1,12 +1,6 @@
 name: Linux
 
-on:
-  push:
-    branches:
-      - main
-  pull_request:
-    branches:
-      - '*'
+on: [push]
```
Comment (bradfitz) [inline .github/workflows/linux.yml:3]: why this change?

File: .github/workflows/xe-experimental-vm-test.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+name: "[experimental] VM-based integration testing, ping @Xe if it fails"
```
Comment (bradfitz) [inline .github/workflows/xe-experimental-vm-test.yml:1]: long name is long :)

Everything we have is experimental to some degree, and people can ping #eng if anythign fails they don't understand, so just "integration-vms" is a good short name.

File: .github/workflows/linux.yml (hunk 2 above)
Comment (Xe) [reply .github/workflows/linux.yml:3]: I think that was on accident, reverting

File: .github/workflows/xe-experimental-vm-test.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - 'main'
+  release:
+    types: [ created ]
+  schedule:
+    - cron: '00 6,18 * * *'
```
Comment (bradfitz) [inline .github/workflows/xe-experimental-vm-test.yml:13]: I forgot this magic syntax. Comment alongside it?

File: .github/workflows/xe-experimental-vm-test.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    runs-on: [ self-hosted, linux, vm_integration_test ]
+
+    if: "!contains(github.event.head_commit.message, '[ci skip]')"
+
+    steps:
+      - name: Checkout Code
+        uses: actions/checkout@v1
+
+      - name: Download VM Images
+        run: go test ./tstest/integration/vms -run-vm-tests -run Download -timeout 60m
```
Comment (bradfitz) [inline .github/workflows/xe-experimental-vm-test.yml:26]: use `=` to make it obvious what what's a flag value... `--run=Download`

File: .github/workflows/xe-experimental-vm-test.yml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        env:
+          XDG_CACHE_HOME: "/var/lib/ghrunner/cache"
+
+      - name: Run VM tests
+        run: go test ./tstest/integration/vms -v -run-vm-tests -run Integration
```
Comment (bradfitz) [inline .github/workflows/xe-experimental-vm-test.yml:31]: should be able to drop `--run` here entirely, right? The Download should be a no-op if it already exists on disk?

File: .github/workflows/xe-experimental-vm-test.yml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  schedule:
+    - cron: '00 6,18 * * *'
+
+jobs:
+  experimental-linux-vm-test:
+    runs-on: [ self-hosted, linux, vm_integration_test ]
+
+    if: "!contains(github.event.head_commit.message, '[ci skip]')"
+
+    steps:
```
Comment (bradfitz) [inline .github/workflows/xe-experimental-vm-test.yml:21]: Don't you need to force Go 1.16 like Linux does?

File: .github/workflows/xe-experimental-vm-test.yml (hunk 6 above)
Comment (Xe) [reply .github/workflows/xe-experimental-vm-test.yml:31]: Is there a way to do test-specific timeouts? I have them as separate steps so cache warming has a longer timeout.

File: .github/workflows/xe-experimental-vm-test.yml (hunk 7 above)
Comment (Xe) [reply .github/workflows/xe-experimental-vm-test.yml:21]: No, the github runner is configured to have nixpkgs Go already in its path with a NixOS module that looks like this:

```nix
{ lib, config, pkgs, ... }:

{
  services.github-runner = {
    enable = true;

    url = "https://github.com/tailscale/tailscale";
    name = "xe-tstest-integration-vms";
    replace = true;
    tokenFile = "/run/decrypted/ts-oss-ghaction-token";
    extraLabels = [ "vm_integration_test" ];
    extraPackages = with pkgs; [ go cdrkit qemu openssh gcc ];
  };

  users.users.ghrunner = {
    createHome = true;
    isSystemUser = true;
    extraGroups = [ "kvm" ];
  };

  systemd.services.github-runner = {
    serviceConfig = {
      DeviceAllow = lib.mkForce [ "/dev/kvm" ];
      DynamicUser = lib.mkForce false;
      PrivateDevices = lib.mkForce false;
      ProtectSystem = lib.mkForce null;
      User = "ghrunner";
      ExtraGroups = [ "kvm" ];
    };
  };
}
```

File: .github/workflows/xe-experimental-vm-test.yml (hunk 4 above)
Comment (Xe) [reply .github/workflows/xe-experimental-vm-test.yml:13]: I added a comment explaining it as well as a link to crontab.guru

File: .github/workflows/xe-experimental-vm-test.yml (hunk 6 above)
Comment (bradfitz) [reply .github/workflows/xe-experimental-vm-test.yml]: You already did Download in the previous Step with a large timeout. If it's already on disk, you can run it again and it'll be quick, without needing a large timeout.

File: .github/workflows/xe-experimental-vm-test.yml (hunk 7 above)
Comment (bradfitz) [reply .github/workflows/xe-experimental-vm-test.yml:21]: That's a little indirect. Check in the nix configs too, then, and add a comment in the github action yml referencing the nix configs, so it's not so implicit?

File: tstest/integration/vms/vms_test.go [hunk 8]
```diff
@@ -360,20 +360,20 @@ func TestVMIntegrationEndToEnd(t *testing.T) {
 
 		{"amazon-linux", "https://cdn.amazonlinux.com/os-images/2.0.20210427.0/kvm/amzn2-kvm-2.0.20210427.0-x86_64.xfs.gpt.qcow2", "6ef9daef32cec69b2d0088626ec96410cd24afc504d57278bbf2f2ba2b7e529b", 512, "yum"},
 		{"arch", "https://mirror.pkgbuild.com/images/v20210515.22945/Arch-Linux-x86_64-cloudimg-20210515.22945.qcow2", "e4077f5ba3c5d545478f64834bc4852f9f7a2e05950fce8ecd0df84193162a27", 512, "pacman"},
-		{"centos-7", "https://cloud.centos.org/centos/7/images/CentOS-7-x86_64-GenericCloud.qcow2", "1db30c9c272fb37b00111b93dcebff16c278384755bdbe158559e9c240b73b80", 512, "yum"},
```
Comment (bradfitz) [inline tstest/integration/vms/vms_test.go:363]: again, move this file's changes out of this PR

File: .github/workflows/xe-experimental-vm-test.yml (hunk 6 above)
Comment (Xe) [reply .github/workflows/xe-experimental-vm-test.yml]: Oh, true.

File: .github/workflows/xe-experimental-vm-test.yml (hunk 7 above)
Comment (Xe) [reply .github/workflows/xe-experimental-vm-test.yml:22]: I added `runner.nix` (a much more commented version of that file) to tstest/integration/vms

**Review by bradfitz — APPROVED** [review]
